### DISCUSSION TITLE
poly data creation from array time delay

### DISCUSSION THREAD

**Discussion by 1939938853** [pyvista/pyvista#1405] [Q&A] [answered]
Hello,

Pyvista is really a great 3D visualization tool! I like it.

I just wonder how can I efficiently get 3D data from other program like Open3D. I want to process data in Open3D and to do the visualization using pyvista. I tried as the following

` 
  
        pcd = open3d.io.read_point_cloud("../../TestData/source.pcd")

        start = time.time()

        point_cloud = pyvista.PolyData(np.asarray(pcd.points))

        print("time elapsed: {} sec".format(time.time() - start))
`
It is a ca 590000 points data. The data transfer from open3d to pyvista  takes 0.43 sec. It is hundred time longer than I expected. Maybe there is memory copy internally? What is the best way?

Further more, is it possible to use GPU?

Thank you!

**Comment by GuillaumeFavelier** [comment]
> I want to process data in Open3D and to do the visualization using pyvista.

Thanks for your interest in `pyvista`. I'm not familiar with `Open3D` but I hope that we can help make it work for you. I don't know the `*.pcd` format but did you consider exporting/converting into a vtk- or pyvista-compatible file format? We also support `meshio` which offers a wide variety of options (they provide a [very useful chart](https://github.com/nschloe/meshio#io-speed) that compares the file format in I/O operations ). If that's the case, it would be as easy as using [`pyvista.read()`](https://docs.pyvista.org/utilities/utilities.html?highlight=read#pyvista.read). 

> It is hundred time longer than I expected. Maybe there is memory copy internally?

We avoid deep copy as much as possible but I don't exclude this possibility.

> Further more, is it possible to use GPU?

I'm not sure to understand your question here. The internal algorithms will run on the CPU but the rendering will be done on the GPU. Other @pyvista/developers could help clarify this.

**Comment by 1939938853** [comment]
Hello,

 I mean here is that pyvista shows too slow when take data from numpy array.  If pyvista can make zero-copy just make use the numpy pointer?

Thank you

**Comment by GuillaumeFavelier** [comment]
Is there a way that you could share a test dataset? It could be helpful to benchmark all those operations:

* reading the data
* creating the suitable pyvista dataset
* display in the render window

>  If pyvista can make zero-copy just make use the numpy pointer?

There could be a little overhead to turn the raw numpy array into a VTK-friendly dataset? I need to investigate. I don't think there is a general answer since it depends on the type of data and the operations done for example.

**Comment by 1939938853** [comment]
Hi, 

By running the following code. It shows

time duration from numpy array to open3d point 0.0026717185974121094 sec
time duration from numpy array to pyvista point 0.10618400573730469 sec

Maybe there is other better way to import numpy array point to pyvista?


```python
import pyvista as pv
import open3d as o3d
import numpy as np
import  time

x = np.linspace(-3, 3, 401)
mesh_x, mesh_y = np.meshgrid(x, x)
z = np.sinc((np.power(mesh_x, 2) + np.power(mesh_y, 2)))
z_norm = (z - z.min()) / (z.max() - z.min())
xyz = np.zeros((np.size(mesh_x), 3))
xyz[:, 0] = np.reshape(mesh_x, -1)
xyz[:, 1] = np.reshape(mesh_y, -1)
xyz[:, 2] = np.reshape(z_norm, -1)
print('xyz')
print(xyz)

pcd = o3d.geometry.PointCloud()
start=time.time()
pcd.points = o3d.utility.Vector3dVector(xyz)
print("time duration from numpy array to open3d point {}".format(time.time() - start))
#o3d.visualization.draw_geometries(([pcd]))
start=time.time()
pv_points = pv.PolyData(xyz)
print("time duration from numpy array to pyvista point {}".format(time.time() - start))

```

**Comment by akaszynski** [comment, marked as answer]
There's a nasty loop that's triggered that we need to fix.  For now, here:

```python
start=time.time()

pv_points = pv.PolyData()

# create faces array
npoints = xyz.shape[0]
faces = np.empty(npoints*2, dtype=pv.ID_TYPE)
faces[1::2] = np.arange(npoints, dtype=pv.ID_TYPE)

# and just shallow copy it
pv_points.points = xyz
pv_points.faces = faces.reshape(-1, 2)

print("time duration from numpy array to pyvista point {}".format(time.time() - start))

```
Timing is about the same.

```
time duration from numpy array to open3d point 0.0009667873382568359
time duration from numpy array to pyvista point 0.0009324550628662109
```

**Comment by 1939938853** [comment]
Great, Thank you!
